go.sh

```diff
@@ -76,6 +76,15 @@ if [ -n "${CLIENT_BODY_BUFFER_SIZE:-}" ]; then
     msg "Setting '${UPLOAD_SETTING};'"
 fi
 
+if [ -n "${ENABLE_BIG_BUFFERS:-}" ]; then
+cat > /etc/nginx/conf/nginx_big_buffers.conf <<-EOF_BIGBUFFERS
+    # Proxy Buffer Size Increase
+    proxy_buffer_size 16k;
+    proxy_busy_buffers_size 48k;
+    proxy_buffers 16 16k;
```

Comment:
curious is the size `16k` here controlling the same property that the value from `proxy_buffer_size` is, like its another way of setting the same thing without also changing the number of buffers? The docs don't seem clear but it kind of looks like it?
http://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_buffering

Reply:
These are different settings. They are similarly named, but they behave differently. The first one is the most important - increasing this is the thing that will address the error we are trying to solve regarding "upstream headers too big". It has to be at least as large as the largest header we might get from the proxy.

The latter option, proxy_buffers, defines how much memory should be allocated to buffering data from the proxy - if it's too low, then the buffers will be written to disk, impacting performance.

[This article](https://www.getpagespeed.com/server-setup/nginx/tuning-proxy_buffer_size-in-nginx) goes into some nice detail about what the settings mean and how to tweak them.

Reply:
That's a brilliant article (assuming its factually correct). Our current `proxy_buffer_size` is 4k based our page size
```
bash-5.0# getconf PAGESIZE
4096
```
Quadrupling that to 16k doesn't seem unreasonable. I'm drawn to the point where it says
> The proxy_buffer_size is the only directive that needs tuning in order to solve the upstream sent too big header while reading response header from upstream error.

I see the bit about `proxy_busy_buffer_size`
> Essentially, its minimum allowed value equals to the value you’d set for proxy_buffer_size.

...and that NGINX will set a default that meets this criteria

> The default value for proxy_busy_buffers_size is calculated dynamically during configuration evaluation state. It’s set to twice the size of either:

>one of the buffers set by proxy_buffers, if those buffers are smallest in comparison to proxy_buffer_size one
value of proxy_buffer_size, otherwise

Curious if there's a reason why you've specifically set `proxy_busy_buffer_size` higher?

Again, curiosity at play, changing `proxy_buffers 16 16k;` doesn't appear necessary to fix the `upstream sent too big header while reading` issue, just curious why this is also being bumped up?

Reply:
Regarding the change to the `proxy_buffers` setting - if we needed to increase the default `proxy_buffer_size` in order to prevent errors from occurring in production, then that means in at least some cases we are processing buffer sizes that are larger than the defaults are set to handle.

We could just leave the others at their defaults, but from what I understand, this would not be optimal for performance as Nginx would have to write to IO when data from the proxy exceeds the buffer size. So if we know we need to increase the buffer size because Nginx is having to handle larger headers, we should also increase the `proxy_buffers` setting to prevent writing to disk becoming necessary.

And as for the proxy_busy_buffer_size, Nginx will determine this by default, based on the other values specified. But apparently, [this calculation is error-prone](https://www.getpagespeed.com/server-setup/nginx/tuning-proxy_buffer_size-in-nginx):
> This automatic configuration is error-prone if it results in the calculated value falling outside the allowed ranges.
To understand the best value for it, we have to touch on proxy_buffers first.

Plus, seeing other user comments, such as:
> When I only define proxy_buffer_size then I get a "proxy_busy_buffers_size" must be less than the size of all "proxy_buffers" minus one buffer warning for nginx -t.

...suggests that even though the Nginx documentation implies that it _should_ calculate a safe setting for this automatically, in reality, it looks like it doesn't.

Therefore, it is probably safer for us to define if we've changed the values that Nginx relies on to determine this value. So whilst we could just take the double of the buffer_size to determine the size of the busy_buffer (32K), I have used the maximum header size (16K) plus two "extra" 16K buffers - hence the result of 48K.